Refactor to use functions and be more efficient on repeated runs.

solarce opened the pull request:
- In order to better isolate some rvm related errors I rewrote things to
  use functions and also added some logic to make it faster when you
  are repeatedly running the script on the same Base VM.

solarce committed:
Refactor to use functions and be more efficient on repeated runs.

- In order to better isolate some rvm related errors I rewrote things to
  use functions and also added some logic to make it faster when you
  are repeatedly running the script on the same Base VM.

emdantrim reviewed (changes requested):
Yay `rvm` optimizations! I think we can just overwrite all the config files each time, though.

solarce committed:
Removing checks for some files

- After further discussion we've decided it makes the most sense
  to always write some files, that may change, until we move to
  something "smarter", like Chef.

emdantrim commented on bootstrap.sh:
```diff
@@ -33,21 +30,34 @@ bootstrap() {
 
-  echo "--- Put hardened sshd config in place"
+passwordless_sudo_setup() {
+  if [[ ! -f /etc/sudoers.d/travis ]]; then
```
I think I'd prefer that this logic was removed as well.

emdantrim commented on bootstrap.sh:
```diff
@@ -65,133 +75,228 @@ PermitRootLogin no
+}
+
+gemrc_setup() {
+  if [[ ! $(grep no-document ~/.gemrc) ]]; then
```
This logic can go too.

emdantrim commented on bootstrap.sh:
```diff
@@ -65,133 +75,228 @@ PermitRootLogin no
+}
+
+disable_scheduled_software_updates() {
+  if [[ ! $(sudo softwareupdate --schedule | grep off) ]]; then
```
I don't trust OS X to not randomly turn this back on after system updates, so it'd be nice to run this every time.

emdantrim reviewed (changes requested):
Just a couple more similar changes to earlier. Otherwise, this is good.

solarce committed:
Remove some additional checks

- Until we port this to something "smarter" like Chef, we prefer to
  accept the "time cost" of doing these steps on every run.

solarce committed:
Merge branch 'master' into sol-rvm-yakshave

emdantrim reviewed (approved):
:sparkles:
